quinn-proto/src/endpoint.rs

```diff
@@ -188,17 +193,46 @@ impl Endpoint {
         //
 
         let addresses = FourTuple { remote, local_ip };
-        if let Some(ch) = self.index.get(&addresses, &first_decode) {
-            return Some(DatagramEvent::ConnectionEvent(
-                ch,
-                ConnectionEvent(ConnectionEventInner::Datagram {
-                    now,
-                    remote: addresses.remote,
-                    ecn,
-                    first_decode,
-                    remaining,
-                }),
-            ));
+        if let Some(route_to) = self.index.get(&addresses, &first_decode) {
+            let event = DatagramConnectionEvent {
+                now,
+                remote: addresses.remote,
+                ecn,
+                first_decode,
+                remaining,
+            };
+            match route_to {
+                RouteDatagramTo::Incoming(incoming_idx) => {
+                    let incoming_buffer = &mut self.incoming_buffers[incoming_idx];
+                    let config = &self.server_config.as_ref().unwrap();
+
+                    if incoming_buffer
+                        .total_bytes
+                        .saturating_add(datagram_len as u64)
+                        < config.max_buffer_bytes_per_incoming
+                        && self
+                            .all_incoming_buffers_total_bytes
+                            .saturating_add(datagram_len as u64)
+                            < config.max_buffer_bytes_all_incoming
+                    {
+                        incoming_buffer.datagrams.push(event);
+                        incoming_buffer.total_bytes = incoming_buffer
+                            .total_bytes
+                            .saturating_add(datagram_len as u64);
```

Comment:
`saturating_add` here and below is superfluous, because any summation that would wrap would have produced `u64::MAX` in the condition above and necessarily failed a `<` test. I wouldn't mind keeping it for symmetry, but just using `+= datagram_len as u64` is more readable and avoids bugs like below.

Reply:
Reworked this logic as described in comment above. Problem should be fixed.